### PR TITLE
119 improve api

### PR DESCRIPTION
**Pull request by maho3**
This PR implements numerous structural changes to the configurations of the Data/Inference/Validation stages, without much changes to how they work on the backend. The general theme of this PR is to add various quality-of-life improvements that make it easier to understand the configurations and use ltu-ili for distributed testing. Here's a summary of what has changed.

Removed unnecessary configurations:
  - Removed the need to specify `backend` in the ValidationRunner. This is now gathered automatically from the imports.
  - Removed the need to specify `n_walkers` in DelfiRunner (inference stage). It didn't make sense to specify `n_walkers` before the sampling stage, and it is redefined anyway in the EmceeSampler during validation.
  - Removed the need to specify `n_data` and `n_params` in the DelfiRunner configuration. This info is now gathered automatically from the provided dataloader.
  - Removed the `n_data` configuration of the FCN network. It was simply the dimensionality of the output of the network. It can now be configured as the last entry in `n_hidden`
  - Removed the `inference_class` argument of the inference Runner objects and replaced it with `engine`.  Previously, users had to load an engine object from file before passing it into the Inference runners (e.g. `inference_class = sbi.inference.SNPE`. Now users only need to specify a string (e.g. `engine='NPE'` or NLE/NRE/SNPE etc) and we figure out what object `inference_class` dependent on the backend.
  - All path-like arguments (e.g. `out_dir`, `in_dir`, etc) can be specified with `str` inputs instead of `pathlib.Path`. We convert them all to `Path`'s on the backend.

Renamed things to make them more obvious:
  - Changed `output_path` -> `out_dir` and `posterior_path` -> `posterior_file` to match the naming conventions in loaders.py. Now, the ValidationRunner will only load the posterior from `out_dir / posterior_path` and subsequently save metrics in `out_dir`. I didn't see a reason why one would want to save metrics in a different directory from their posterior.
  - Similar to the last change, `output_path` -> `out_dir` in the Inference stages.
  - All stages can now be loaded from their top-level category. E.g.
  ```
  from ili.dataloaders import StaticNumpyLoader
  from ili.inference import SBIRunner
  from ili.validation import ValidationRunner
  ```
  Note, The old imports (e.g. `from ili.inference.runner_sbi import SBIRunner` still work for backwards compatibility.

Conformed the multiple Inference backends to a universal configuration:
  - Added `InferenceRunner` in ili/inference/runner.py to act as a universal engine for the Inference stage. It's configuration takes `backend` and `engine` as parameters, and then uses these to go off and find the write object for your inference. For example, `backend='sbi'` and `engine='NLE'` will load the configuration in `SBIRunner`, whereas `backend='sbi'` and `engine='SNLE'` will load it in `SBIRunnerSequential`. This new configuration should be preferred, but it does not affect the way we previously specified SBIRunner/SBIRunnerSequential/DelfiRunner etc, so it should be backwards compatible. Adds appropriate tests.
  - Added ili/utils/ndes_pt.py and ili/utils/ndes_tf.py which provide the `load_nde_sbi` and `load_nde_pydelfi` which provide utilities for loading neural architectures from given configurations for each backend. This unifies the configuration with which we specify neural architectures between sbi and pydelfi, making things considerably simpler. These functions were used to refactor `SBIRunner` and `DelfiWrapper`. Adds appropriate tests.
  - Added the ability to overload a loaded config file in each `from_config` function of the Data/Inference/Validation stages. This can be very useful if you want to iterate (in python) over many small changes to a configuration file. For example, one could change the training rate and the `out_dir` to save many models to different directories, each of which trained with a different training rate. Addresses #119 
  - `DelfiWrapper`'s and `EmceeSampler`'s sampler functions were fixed to only produce `nsteps` samples, matching the behavior pyro samplers. Previously, they were producing `nsteps*num_chains` samples.
  - DelfiRunner now saves a summary of train/validation loss, following SBIRunner.

Miscellaneous:
  - Updated the jupyter notebook example with the new interface.
  - Updated CONFIG.md accordingly.
  - Added `_BaseRunner` in ili/inference/base.py as a template for sbi/pydelfi runners

Note a lot of files have changed, because we needed to propagate the naming changes to all the examples and tests. I'm happy to walk the reviewer through the listed changes.

### CODE REVIEW

**Comment by CompiledAtBirth** [comment]
Currently running tests on a duplicate of this branch (seq_test to use a more recent torch than the setup.cfg).

**Review by DeaglanBartlett — CHANGES_REQUESTED** [review]
Really excellent work! Apart from a couple of minor comments I left, please can you add tests for the following, which are missing (according to Codecov):

- [x] - `utils.ndes_pt.load_nde_sbi()` should raise an error if the `model` is not one of `['mdn', 'maf', 'nsf', 'made']`
- [x] - An error is raised if neither `x` nor `x_obs` is provided for `PlotSinglePosterior`
- [ ] - The numpy seed is updated in `PlotSinglePosterior` if `x_obs` is `None` and a seed is provided
- [x] - Check samples can be saved from `PlotSinglePosterior`
- [ ] - Check `ValidationRunner.load_posterior_sbi()` works even if the posterior does not have a `name` attribute
- [x] - Check an error is raised if an invalid backend if provided to `InferenceRunner._parse_runner()`
- [x] - Check `DelfiRunner` still works even if you incorrectly use a device which isn't `cpu`
- [x] - Test loading of `engine_kwargs` for `DelfiRunner`
- [x] - Check for `ImportError` if the wrong engine is supplied to `SBIRunner._setup_engine()`
- [x] - Check for errors if one attempts to use `SBIRunnerSequential` with a loader which doesn't have attributes get_obs_data or simulate

**Comment by DeaglanBartlett** [comment]
From @CompiledAtBirth on slack:

> For the SBISimulator loader specifically, the simulator function has to output a torch.Tensor of torch.Size((1,n)). The case is easy to handle once the user knows his simulator function must return that.

To avoid this mistake, please can you make sure it is well documented in the code and in the tutorial? We might as well do it in this PR as it is a minor thing to add

File: ili/validation/runner.py [hunk 1]
```diff
@@ -8,79 +8,94 @@
...
-                config["posterior_path"])
+                out_dir / config["posterior_file"])
             signatures = posterior_ensemble.signatures
-        elif backend == 'pydelfi':
-            posterior_ensemble = DelfiWrapper.load_engine(config["meta_path"])
-            signatures = [""]*posterior_ensemble.num_components
+        elif interface == 'tensorflow':
+            posterior_ensemble = DelfiWrapper.load_engine(
+                out_dir / config["posterior_file"])
+            signatures = [""]*posterior_ensemble.num_components  # TODO: fix
```
[the remark this answers is not in the document]
Comment (maho3) [reply ili/validation/runner.py:95]: This was meant to add signature functionality to the individual pydelfi models. However, this is a bit tricky as you need to redo the DelfiRunner save/load function again. Best saved for a future revision. I made a note of this in-line and removed the todo.

File: tests/test_sbi.py [hunk 2]
```diff
@@ -420,7 +419,10 @@ def simulator(params):
     x0 = simulator(theta0)
     np.save('toy/thetaobs.npy', theta0[0])
     np.save('toy/xobs.npy', x0[0])
-    
+
+    np.save('toy/theta.npy', theta)
```
[the remark this answers is not in the document]
Comment (maho3) [reply tests/test_sbi.py:423]: Removed. This was likely a merge issue.

**Comment by maho3** [comment]
> From @CompiledAtBirth on slack:
> 
> > For the SBISimulator loader specifically, the simulator function has to output a torch.Tensor of torch.Size((1,n)). The case is easy to handle once the user knows his simulator function must return that.
> 
> To avoid this mistake, please can you make sure it is well documented in the code and in the tutorial? We might as well do it in this PR as it is a minor thing to add

Added comment in the code and tutorial.

**Comment by maho3** [comment]
> The numpy seed is updated in PlotSinglePosterior if x_obs is None and a seed is provided

This is rather difficult to test because we don't return what x_obs we choose, and the used np seed is difficult to access. I think this test isn't necessary, given the code is pretty clear as to what its doing.

**Comment by maho3** [comment]
> Check ValidationRunner.load_posterior_sbi() works even if the posterior does not have a name attribute

I just removed this functionality of `load_posterior_sbi`, because there should never be a case wherein an sbi posterior doesn't have a `name`. It is given `""` by default.

**Review by DeaglanBartlett — APPROVED** [review]
